wallet: allow psbtbumpfee to work with txs with external inputs

achow101 opened the pull request:
This PR allows `psbtbumpfee` to return a PSBT for transactions that contain external inputs. This does not work for bumping in the GUI nor `bumpfee` because these need private keys available to sign and send the transaction. But `psbtbumpfee` returns a psbt, so it is fine to not be able to sign.

In order to correctly estimate the size of the inputs for coin selection, the fee bumper will use the size of the inputs of the transaction being bumped. Because the sizes of signatures are not guaranteed, for external inputs, the fee bumper will verify the scripts with a special SignatureChecker which will compute the weight of all of the signatures in that input, and compute their weights if those signatures were maximally sized. This allows the fee bumper to obtain a max size estimate for each external input.

Builds on #23201 as it relies on the ability to pass weights in to coin selection.

Closes #23189

achow101 committed:
bumpfee: Calculate fee by looking up UTXOs

Instead of calculating the fee by using what is stored in the wallet,
calculate it by looking up the UTXOs.

achow101 committed:
bumpfee: extract weights of external inputs when bumping fee

When bumping the fee of a transaction containing external inputs,
determine the weights of those inputs. Because signatures can have a
variable size, the script is executed with a special SignatureChecker
which will compute the total weight of the signatures in the transaction
and the weight if they were all maximum size signatures. This allows us
to compute the maximum weight of the input for use during coin
selection.

achow101 committed:
bumpfee: be able to bump fee of a tx with external inputs

In some cases, notably psbtbumpfee, it is okay, and potentially desired,
to be able to bump the fee of a transaction which contains external
inputs.

achow101 committed:
test, bumpfee: Check that psbtbumpfee can bump txs with external inputs

prayank23 commented:
Concept ACK

prayank23 commented:
Compiling this branch now as its ready for review. Not sure if I will be able to test everything but this is an interesting feature which should exist in Core. Hoping other devs review this without any club.

t-bast commented:
I would really love to see this land. I can test this end-to-end with eclair whenever it's ready and verify that it lets us bump lightning transactions (I can add tests with various kinds of inputs, confirmed / unconfirmed / unsafe).

How does it impact @Xekyo's work on ancestor aware funding (https://github.com/Xekyo/bitcoin/commits/ancestor-aware-funding)? Ideally the feerate computed here should take into account all unconfirmed ancestors, but maybe Xekyo's work would just land after this PR and fix its feerate computation?

achow101 commented:
> How does it impact @Xekyo's work on ancestor aware funding (https://github.com/Xekyo/bitcoin/commits/ancestor-aware-funding)? Ideally the feerate computed here should take into account all unconfirmed ancestors, but maybe Xekyo's work would just land after this PR and fix its feerate computation?

I believe that work will require some mempool changes that have not been implemented yet, so it is probable that it will still be some time before that is ready for review.

t-bast commented:
Thanks for the quick answer, so I guess @Xekyo's work will simply improve `psbtbumpfee`'s feerate estimation when there are unconfirmed ancestors and can be added after this PR is merged.

ishaanam commented on src/wallet/feebumper.cpp:
```diff
@@ -150,12 +150,64 @@ bool TransactionCanBeBumped(const CWallet& wallet, const uint256& txid)
+            m_sigs_weight += weight;
+            m_sigs_count++;
+            break;
+        case SigVersion::TAPROOT:
```
In "bumpfee: extract weights of external inputs when bumping fee" (a95d41bdf366939e4d3ff6368c384e752d745648)

Do we never need to do this for witness v1 outputs because these signatures are always 64 bytes so there is no need to estimate? If this is the case, would I be correct in saying that CalculateMaximumSignedInputSize() doesn't return -1 for taproot outputs in this case?

achow101 replied:
Yes and yes.

Xekyo replied:
It would have really helped me if you had explained a bit more here.

Xekyo replied:
Wouldn't it be possible that you're creating a replacement transaction for which you have a different spending condition that you can solve than the transaction that you are replacing? E.g. you're using a scriptpath and the original transaction was a keypath spend?

achow101 replied:
> Wouldn't it be possible that you're creating a replacement transaction for which you have a different spending condition that you can solve than the transaction that you are replacing? E.g. you're using a scriptpath and the original transaction was a keypath spend?

Yes, but we can't know that. Either the wallet is the one spending the input and so we are able to estimate the size using whatever the wallet will sign with, or the input is external and we only know the spending path that exists in the transaction to be bumped.

ishaanam reviewed (commented):
Concept ACK, I have a few questions to make sure that I'm understanding how this is being implemented correctly.

achow101 commented on src/wallet/feebumper.cpp:
```diff
@@ -169,13 +167,30 @@ Result CreateRateBumpTransaction(CWallet& wallet, const uint256& txid, const CCo
+    for (const CTxIn& txin : wtx.tx->vin) {
+        const Coin& coin = coins.at(txin.prevout);
+        assert(!coin.out.IsNull());
+        new_coin_control.SelectExternal(txin.prevout, coin.out);
```
There is largely no difference, but since #24649, our pattern is to actually only use `SelectExternal` for external inputs, so I have changed this to do that.

achow101 commented:
Rebased for silent merge conflict.

ishaanam reviewed (commented):
Approach ACK, I just have some minor questions and a nit.

achow101 commented on src/wallet/feebumper.cpp:
```diff
@@ -169,13 +167,34 @@ Result CreateRateBumpTransaction(CWallet& wallet, const uint256& txid, const CCo
+    wallet.chain().findCoins(coins);
+    for (const CTxIn& txin : wtx.tx->vin) {
+        const Coin& coin = coins.at(txin.prevout);
+        assert(!coin.out.IsNull());
```
That is theoretically possible. Changed this to return an error in that case.

achow101 commented on src/wallet/feebumper.cpp:
```diff
@@ -152,6 +153,58 @@ bool TransactionCanBeBumped(const CWallet& wallet, const uint256& txid)
+struct SignatureWeights
+{
+    int m_sigs_count{0};
+    int64_t m_sigs_weight{0};
```
For consistency with `GetMaxSigsWeight`. I've made it private anyways.

achow101 commented on src/wallet/feebumper.cpp:
```diff
@@ -152,6 +153,58 @@ bool TransactionCanBeBumped(const CWallet& wallet, const uint256& txid)
+    int m_sigs_count{0};
+    int64_t m_sigs_weight{0};
+
+    void AddSigWeight(size_t weight, SigVersion sigversion)
```
Done.

Xekyo commented on src/wallet/feebumper.cpp:
```diff
@@ -150,12 +150,66 @@ bool TransactionCanBeBumped(const CWallet& wallet, const uint256& txid)
+            m_sigs_count++;
+            break;
+        case SigVersion::TAPROOT:
+        case SigVersion::TAPSCRIPT:
+            assert(false);
```
I think it might be good to catch all other calls to this function that are not covered by BASE or WITNESS_V0 as not being expected to be handled in this function. This will make this function future proof if other SigVersions get added in the future.

```suggestion
        case SigVersion::TAPSCRIPT:
        default:
            assert(false);
```

sipa replied:
It's actually better to not have a `default:` clause, as it means the compiler can warn about missing clauses, rather than needing to notice through an assertion failure.

achow101 replied:
Yes, the lack of a `default` is intentional.

fanquake replied:
Going to mark this as resolved.

ishaanam commented:
ACK 66a5cb5b4651191bc92438cbc3f02226d250feef

fanquake commented:
@instagibbs want to take a look here given #23189?

t-bast reviewed (approved):
Tested ACK https://github.com/bitcoin/bitcoin/pull/23202/commits/66a5cb5b4651191bc92438cbc3f02226d250feef, thanks @achow101!

I integrated this with lightning to bump the fees of anchor transactions and htlc transactions and was able to bump both kinds of transactions correctly.

There was one caveat though for the anchor transaction, that should probably be fixed in a separate PR. This is a transaction that is only used to CPFP its parent transaction, so it has a single output sending to a change address. Doing `psbtbumpfee` yields the following error: `Unable to create transaction. Transaction must have at least one recipient (code: -4)`. Browsing through the code it looks like `bitcoind` doesn't count an output sending to a change address as a "real" output in `CreateTransaction` (https://github.com/bitcoin/bitcoin/blob/27724c23f72b17142f95e082e23eee26e93791f4/src/wallet/spend.cpp#L1067). I can work around this, but a follow-up PR to make this check more lenient would IMO be useful.

instagibbs commented on src/wallet/interfaces.cpp:
```diff
@@ -289,7 +289,7 @@ class WalletImpl : public Wallet
         CMutableTransaction& mtx) override
     {
-        return feebumper::CreateRateBumpTransaction(*m_wallet.get(), txid, coin_control, errors, old_fee, new_fee, mtx) == feebumper::Result::OK;
+        return feebumper::CreateRateBumpTransaction(*m_wallet.get(), txid, coin_control, errors, old_fee, new_fee, mtx, /* require_mine= */ true) == feebumper::Result::OK;
```
Might want some motivation somewhere(header declaration?) when `require_mine` can be false or true.

achow101 replied:
added docs

achow101 committed:
wallet, tests: Test bumpfee's max input weight calculation

achow101 committed:
Make DUMMY_CHECKER availble outside of script/sign.cpp

achow101 commented on src/wallet/feebumper.cpp:
```diff
@@ -187,6 +243,31 @@ Result CreateRateBumpTransaction(CWallet& wallet, const uint256& txid, const CCo
+            // signature sizes and replace them with the max sized signature.
+            // In order to do this, we verify the script with a special SignatureChecker which
+            // will observe the signatures verified and record their sizes.
+            SignatureWeights weights;
```
Added a test.

achow101 commented on src/wallet/feebumper.cpp:
```diff
@@ -187,6 +243,31 @@ Result CreateRateBumpTransaction(CWallet& wallet, const uint256& txid, const CCo
+            SignatureWeightChecker size_checker(weights, tx_checker);
+            VerifyScript(txin.scriptSig, coin.out.scriptPubKey, &txin.scriptWitness, STANDARD_SCRIPT_VERIFY_FLAGS, size_checker);
+            // Add the difference between max and current to input_weight so that it represents the largest the input could be
+            input_weight += weights.GetMaxSigsWeight() - weights.GetSigsWeight();
```
Done.

achow101 commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1045,7 +1045,7 @@ static RPCHelpMan bumpfee_helper(std::string method_name)
     feebumper::Result res;
     // Targeting feerate bump.
-    res = feebumper::CreateRateBumpTransaction(*pwallet, hash, coin_control, errors, old_fee, new_fee, mtx);
+    res = feebumper::CreateRateBumpTransaction(*pwallet, hash, coin_control, errors, old_fee, new_fee, mtx, !want_psbt);
```
done

achow101 commented:
@t-bast @ishaanam @instagibbs @Xekyo Re-review?

t-bast commented:
There seems to be an issue when I provide the `fee_rate` parameter, I'm getting the following (seemingly unrelated) error:

```
bitcoind: wallet/wallet.cpp:1540: bool wallet::FillInputToWeight(CTxIn&, int64_t): Assertion `txin.scriptWitness.IsNull()' failed.
```

I was calling `psbtbumpfee` on the following regtest tx:

```
02000000000102b72c3d62593110e0e340e15363b9d12dbe79af9fb9aa2087ba306436d5a51bca0200000000010000003182d6341b973d2f83b9b07a2075a6501e9753831df26427cbe0de6c601b43000100000000fdffffff02bb08000000000000220020352a6494d2541ba95998c1f2881bce71cb4fe420ebda7cba45f66d88222e4ba02858eb0200000000225120996aa742c86fe13934f855a461a3ac23acfca489369a3b80c8829ba5d122cbc005004730440220309709721d51a931c9984ae02c4f6cefc72d947a61b4d46d39f891f0ff24f92402202fd39201ccfa4e201e5f4c6a72fcb248ac985b21946dc07c2d1ccff3f30e45ed83473044022022fe0c3dd42c641e9cd4756360165597333ef24b1aebbe6f1a7f66e7e2288403022040cffadb335097ab36036d18ee18178520bc92b8bac894c430961242fe3e82880120fa64d24aac82ae7f180cd3fb75391a1be94913130d514b75d3d2001018f14f968d76a91458ea2a44f040121f04384be65d29024be6be21a18763ac672102a1af0b6995a260ad47f9438ae5525701c8bdafd3792b8418376384466f1e55777c8201208763a914a451fab97f394a6bb0e74f22992c227e28e62b2f88527c2103faf880218ed6f87580e99a4760857726a74adf878be803567f7babea4166c0a052ae6775022701b175ac6851b2756801401f5c05b6fc4a49e8b5bf4557753e9d861b62b3509bf82c497698328b519bd1f29360b89c095e22b4e1276d8801fe515ed15f46edd8ab74416b241fefef33927d00000000
```

Its second input is a wallet input and the second output is the change output, and the amount should allow bumping to the feerate I'm expecting.

achow101 committed:
bumpfee: Clear scriptSigs and scriptWitnesses before calculated max size

The max size calculation expects some inputs to have empty scriptSigs
and witnesses, so we need to clear these before doing that calculation.

achow101 commented:
> There seems to be an issue when I provide the `fee_rate` parameter, I'm getting the following (seemingly unrelated) error:

Good catch. I've added a fix for that and a test case.

instagibbs commented on src/wallet/feebumper.cpp:
```diff
@@ -242,7 +242,13 @@ Result CreateRateBumpTransaction(CWallet& wallet, const uint256& txid, const CCo
         // The user provided a feeRate argument.
         // We calculate this here to avoid compiler warning on the cs_wallet lock
-        const int64_t maxTxSize{CalculateMaximumSignedTxSize(*wtx.tx, &wallet, &new_coin_control).vsize};
+        // We need to make a temporary transaction with no inputs witnesses as the dummy signer expects them to be empty for some inputs
```
which inputs?

ishaanam replied:
I think inputs for which we have set the weight using `coin_control.SetInputWeight()`, as we do with external inputs here.

achow101 replied:
Clarified for that it's for external inputs.

t-bast reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/23202/commits/bea4313459cb07ae8ba119321e9177d7da0a3bed, I tried to bump lightning htlc transactions and everything seems to work fine :+1:

instagibbs commented:
ACK https://github.com/bitcoin/bitcoin/pull/23202/commits/bea4313459cb07ae8ba119321e9177d7da0a3bed

furszy reviewed (commented):
Needs to be rebased, b3d33efa was included in #25679.

achow101 commented:
> Needs to be rebased, [b3d33ef](https://github.com/bitcoin/bitcoin/commit/b3d33efab794000c6159dea1cba868160cf12b99) was already included in #25679.

Rebased

ishaanam commented on src/wallet/test/feebumper_tests.cpp:
```diff
@@ -0,0 +1,50 @@
+
+BOOST_AUTO_TEST_CASE(external_max_weight_test)
+{
+    CheckMaxWeightComputation("453042021f03c8957c5ce12940ee6e3333ecc3f633d9a1ac53a55b3ce0351c617fa96abe021f0dccdcce3ef45a63998be9ec748b561baf077b8e862941d0cd5ec08f5afe68012102fccfeb395f0ecd3a77e7bc31c3bc61dc987418b18e395d441057b42ca043f22c", {}, "76a914f60dcfd3392b28adc7662669603641f578eed72d88ac", 593);
+    CheckMaxWeightComputation("160014001dca1b22c599b5a56a87c78417ad2ff39552f1", {"3042021f5443c58eaf45f3e5ef46f8516f966b334a7d497cedda4edb2b9fad57c90c3b021f63a77cb56cde848e2e2dd20b487eec2f53101f634193786083f60b4d23a82301", "026cfe86116f161057deb240201d6b82ebd4f161e0200d63dc9aca65a1d6b38bb7"}, "a9147c8ab5ad7708b97ccb6b483d57aba48ee85214df87", 364);
+    CheckMaxWeightComputation("", {"3042021f0f8906f0394979d5b737134773e5b88bf036c7d63542301d600ab677ba5a59021f0e9fe07e62c113045fa1c1532e2914720e8854d189c4f5b8c88f57956b704401", "0359edba11ed1a0568094a6296a16c4d5ee4c8cfe2f5e2e6826871b5ecf8188f79"}, "00149961a78658030cc824af4c54fbf5294bec0cabdd", 272);
+    CheckMaxWeightComputation("", {"3042021f5c4c29e6b686aae5b6d0751e90208592ea96d26bc81d78b0d3871a94a21fa8021f74dc2f971e438ccece8699c8fd15704c41df219ab37b63264f2147d15c34d801", "01", "6321024cf55e52ec8af7866617dc4e7ff8433758e98799906d80e066c6f32033f685f967029000b275210214827893e2dcbe4ad6c20bd743288edad21100404eb7f52ccd6062fd0e7808f268ac"}, "002089e84892873c679b1129edea246e484fd914c2601f776d4f2f4a ...
```
It would be nice to have some comments detailing what kinds of scripts are tested here, it could save people time and it would be easier to tell what there is coverage for at a glance to add more tests in the future.

achow101 replied:
Done

ishaanam commented on test/functional/wallet_bumpfee.py:
```diff
@@ -250,8 +250,27 @@ def test_notmine_bumpfee_fails(self, rbf_node, peer_node, dest_address):
     signedtx = peer_node.signrawtransactionwithwallet(signedtx["hex"])
     rbfid = rbf_node.sendrawtransaction(signedtx["hex"])
+    entry = rbf_node.getmempoolentry(rbfid)
+    old_fee = entry["fees"]["base"]
+    old_feerate = int(entry["fees"]["base"] / entry["vsize"] * Decimal(1e8))
```
nit:
```suggestion
    old_fee = entry["fees"]["base"]
    old_feerate = int(old_fee / entry["vsize"] * Decimal(1e8))

achow101 replied:
Done

ishaanam reviewed (commented):
reACK b1f8fa37028e5947aa250dae8700144947e17986

ishaanam commented:
reACK c3b099ace031758cafeec08c38bedbf717d6b7fe

t-bast reviewed (approved):
Re-ran my tests agains https://github.com/bitcoin/bitcoin/pull/23202/commits/c3b099ace031758cafeec08c38bedbf717d6b7fe, ACK
